### PR TITLE
Issue #6971: Added implementation for NoArrayTrailingCommaCheck

### PR DESCRIPTION
**Pull request by sd1998**
Fix for issue #6971.

Contribution PR: https://github.com/checkstyle/contribution/pull/436

Regression report:
https://sd1998.github.io/checkstyle-regression/Fix-6971/index.html

Base config:
https://sd1998.github.io/checkstyle-regression/Fix-6971/base_config.xml

Patch config:
https://sd1998.github.io/checkstyle-regression/Fix-6971/patch_config.xml

Report repo:
https://github.com/sd1998/checkstyle-regression/tree/master/Fix-6971

### CODE REVIEW

**Comment by sd1998** [comment]
@romani I have added an implementation of NoArrayTrailingCommaCheck

**Comment by rnveach** [comment]
@sd1998 Thanks. It is best we try to finish the already opened PRs. One is waiting on CI to pass.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ * };
+ * String[] foo2 = { "FOO", "BAR", }; // violation
+ * String[] foo3 = {
+ *   "FOO", // OK
+ *   "BAR" // OK
+ * };
+ * String[] foo4 = { "FOO", "BAR" }; // OK
+ * </pre>
+ *
+ * @since 3.2
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java]: it should be 8.XX

File: pom.xml [hunk 2]
```diff
@@ -1142,6 +1142,7 @@
         <artifactId>maven-surefire-plugin</artifactId>
         <version>2.22.2</version>
         <configuration>
+          <forkMode>never</forkMode>
```
Comment (romani) [inline pom.xml]: please revert  or exlain

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages.properties [hunk 3]
```diff
@@ -40,6 +40,7 @@ multiple.variable.declarations.comma=Each variable declaration must be in its ow
 nested.for.depth=Nested for depth is {0,number,integer} (max allowed is {1,number,integer}).
 nested.if.depth=Nested if-else depth is {0,number,integer} (max allowed is {1,number,integer}).
 nested.try.depth=Nested try depth is {0,number,integer} (max allowed is {1,number,integer}).
+no.array.trailing.comma=Array should not contain trailing comma.
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages.properties]: plase use google translator to provide transalations to other languages

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java (hunk 1 above)
Comment (sd1998) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java]: @romani I have made the change.

File: pom.xml (hunk 2 above)
Comment (sd1998) [reply pom.xml]: This one as well.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages.properties (hunk 3 above)
Comment (sd1998) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages.properties]: This one too.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.ARRAY_INIT};
+    }
+
+    @Override
+    public void visitToken(DetailAST arrayInit) {
+        DetailAST rcurly = arrayInit.findFirstToken(TokenTypes.RCURLY);
+        DetailAST previousSibling = rcurly.getPreviousSibling();
+        if (previousSibling != null && previousSibling.getType() == TokenTypes.COMMA) {
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java]: just curious - in what cases `previousSibling` can be null here?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/noarraytrailingcomma/InputNoArrayTrailingComma.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.noarraytrailingcomma;
+
+public class InputNoArrayTrailingComma {
```
Comment (strkkk) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/noarraytrailingcomma/InputNoArrayTrailingComma.java]: I suggest to add test cases with one and zero elements in array

File: src/xdocs/config_coding.xml [hunk 6]
```diff
@@ -3389,6 +3389,86 @@ for (String line: lines) {
       </subsection>
     </section>
 
+    <section name="NoArrayTrailingComma">
+      <p>Since Checkstyle 8.26</p>
```
Comment (strkkk) [inline src/xdocs/config_coding.xml]: you should change it to next version, which is 8.27

**Comment by romani** [comment]
@sd1998 , CI need to be green to let me merge, 
https://travis-ci.org/checkstyle/checkstyle/jobs/611305595#L1205 do you have problems to resolve this ?
All Checks need to be active in our config, this Check with ignore severity, as it confly with another Check.

Only one CI failure is expected - about new Check presence in contribution config, you did https://github.com/checkstyle/contribution/pull/417

**Comment by romani** [comment]
@sd1998 , please rebase and resolve conflict.

please add xpath tests.

**Comment by romani** [comment]
@sd1998 , ping.

**Comment by sd1998** [comment]
@romani `runVerifications` method in `AbstractXpathTestSupport.java` tries to extract the row number as well as the column number but in `NoArrayTrailingCommaCheck.java` the error message only comprises of the row no. (so that it is consistent with what is done in `ArrayTrailingCommaCheck.java`) should I modify the method to extract only the row numbers incase the error message does not contain column numbers?

**Comment by romani** [comment]
yes, looks like we have NO support in Xpath https://github.com/checkstyle/checkstyle/blob/master/src/it/java/org/checkstyle/suppressionxpathfilter/AbstractXpathTestSupport.java#L55 for Checks that report only line number.

It is kind of reasonable as xpath imply detection of exact AST where violation is happening.
Lets report violation on certain line and column. 
Extra comma is definitely has line and column so you will place violation on exact symbol that Check do not like - that is good.

**Comment by rnveach** [comment]
> yes, looks like we have NO support in Xpath AbstractXpathTestSupport for Checks that report only line number.

See https://github.com/checkstyle/checkstyle/issues/5777

**Comment by sd1998** [comment]
@romani I have made all the changes.

**Comment by romani** [comment]
upps .... I did not notice your comment and pushed all changes myself to your remote branch with force.

Please push one more time from your local. Looks like I did the same code changes as you..... I thought that you abandoned your PR.

**Comment by sd1998** [comment]
@romani Done.

File: config/checkstyle_checks.xml [hunk 7]
```diff
@@ -365,6 +365,9 @@
       <property name="max" value="3"/>
     </module>
     <module name="NestedTryDepth"/>
+    <module name="NoArrayTrailingComma">
+      <property name="severity" value="ignore"/>
```
Comment (romani) [inline config/checkstyle_checks.xml]: add above property `<!-- This Check is conflicting with ArrayTrailingComma -->`

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionNoArrayTrailingCommaTest.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+        final List<String> expectedXpathQueries = Collections.singletonList(
+            "/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionNoArrayTrailingComma']]"
+                    + "/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='t1']]/ASSIGN/EXPR"
+                    + "/LITERAL_NEW/ARRAY_INIT/COMMA[4]"
+        );
+
+        runVerifications(moduleConfig, fileToProcess, expectedViolation,
+                expectedXpathQueries);
+    }
```
Comment (romani) [inline src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionNoArrayTrailingCommaTest.java]: please add one more test - https://github.com/romani/checkstyle/commit/74607be0fa6c9df222e5ba77318944adbe5037c8#diff-d48569ac60d48977f90f741b89b997eeR64

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+ * };
+ * String[] foo2 = { "FOO", "BAR", }; // violation
+ * String[] foo3 = {
+ *   "FOO", // OK
+ *   "BAR" // OK
+ * };
+ * String[] foo4 = { "FOO", "BAR" }; // OK
+ * </pre>
+ *
+ * @since 8.27
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java]: 8.28

File: src/xdocs/config_coding.xml [hunk 10]
```diff
@@ -3431,6 +3431,86 @@ for (String line: lines) {
       </subsection>
     </section>
 
+    <section name="NoArrayTrailingComma">
+      <p>Since Checkstyle 8.27</p>
```
Comment (romani) [inline src/xdocs/config_coding.xml]: 8.28

Comment (sd1998) [reply src/xdocs/config_coding.xml]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java (hunk 9 above)
Comment (sd1998) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoArrayTrailingCommaCheck.java]: Done

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionNoArrayTrailingCommaTest.java (hunk 8 above)
Comment (sd1998) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionNoArrayTrailingCommaTest.java]: Done

File: config/checkstyle_checks.xml (hunk 7 above)
Comment (sd1998) [reply config/checkstyle_checks.xml]: Done

**Comment by romani** [comment]
Please satisfy spell checker by whitelisting of word https://travis-ci.org/checkstyle/checkstyle/jobs/629058068#L281

File: src/xdocs/config_coding.xml [hunk 11]
```diff
@@ -3431,6 +3431,86 @@ for (String line: lines) {
       </subsection>
     </section>
 
+    <section name="NoArrayTrailingComma">
+      <p>Since Checkstyle 8.28</p>
+       <subsection name="Description" id="NoArrayTrailingComma_Description">
+         <p>
+           Checks that array initialization do not contain a trailing comma.
+         </p>
```
Comment (pbludov) [inline src/xdocs/config_coding.xml]: There should be a short "rationale" section.
For example:

> Rationale: JLS allows trailing commas in arrays and enumerations, but does not allow them in other locations. To unify the coding style, the use of trailing commas should be prohibited.

Comment (romani) [reply src/xdocs/config_coding.xml]: Agree

Comment (sd1998) [reply src/xdocs/config_coding.xml]: Done

**Comment by sd1998** [comment]
@romani I have updated the regression report.

**Comment by rnveach** [comment]
> Contribution PR: checkstyle/contribution#417

CI failed to notice that PR is not for checks-nonjavadoc-error.xml .

**Comment by romani** [comment]
validation is not ideal ....
message is pointing to correct files - https://github.com/checkstyle/checkstyle/blob/master/.ci/travis/travis.sh#L509
so it is just human mistake, confusion file will be removed at https://github.com/checkstyle/contribution/pull/435
